docs/writing_rules/README.md

```diff
@@ -0,0 +1,272 @@
+> *missing* authn/authz.
+
+These are a few examples, but in summary, the first step in writing a new rule
+is determining:
+
+> What is an interesting aspect of my code that I want to find?
+
+### 2. Determine What That Concretely Looks like in Code
+
+After you have a rough idea of what you want to find, the next step is to
+identify what that concretely looks like in code. The more specific the better.
+
+What does "*good*" or "*safe*" code look like?
+
+For example, after reading some internal code and reading API docs, you find that for the Python `subprocess` module, `call()` and other methods are more dangerous when passed the argument `shell=True`, so you decide:
+
+> I want to find all calls to `subprocess.call()` in which one of the keyword arguments is `shell=True`.
+
+Or perhaps you're auditing Java Spring applications and you want to find all routes that do not perform authorization checks. You review some example routes in your internal repos and see code like this:
+
+```java
+@Controller
+@RequestMapping("/api/")
+public class AcmeController {
+
+    @RequestMapping(method = RequestMethod.POST)
+    @Authorize(Permissions.ADMIN)
+    @ResponseBody
+    public ResponseEntity<Map<String, Object>> createProfile() {
+        return new ResponseEntity<>(result, HttpStatus.OK);
+    }
+
+    @RequestMapping(method = RequestMethod.GET)
+    @ResponseBody
+    public ResponseEntity<Map<String, Object>> showResults() {
+        return new ResponseEntity<>(result, HttpStatus.OK);
+    }
+}
+```
+
+So you determine:
+
+> I want to find all routes, which are methods with the the `@RequestMapping` annotation, that **do not** also have an `@Authorize` annotation.
+
+The key at this stage is to go from an *idea* to one or more *concrete example code snippets* that demonstrate secure and/or insecure examples of what you're looking for.
+
+### 3. Create a Test File, Write an Initial Rule
+
+Now the fun part begins!
+
+Create a YAML file in your current working directory, for example, the project root of a repo you intend to scan. You can name the YAML file whatever you want, but Semgrep will look for rules defined in a file named `.semgrep.yml` by default.
+
+```yaml
+rules:
+  - id: my-pattern-name
+    pattern: |
+      TODO
+    message: "Some message to display to the user"
+    languages: [python]
+    severity: ERROR
+```
+
+Create an example test file, containing snippets of code that **should** and **should not** match.
+
+Fill out the `TODO` in the above YAML file with a Semgrep pattern, using various [pattern features](../pattern-features.md) and [additional pattern clauses](../configuration-files.md) as needed.
+
+For additional help, see [these docs](examples.md) for walkthrough rule writing examples, or an interactive, example-based tutorial here: https://semgrep.live/learn.
+
+Check that the rule you're writing matches the example file by running:
+
+```bash
+$ semgrep example_file.py
```

Comment:
I haven't seen these methods of invocation side-by-side before. My first thought was "why do they need to be run together?" Could we choose one? It's probably more beneficial to be explicit (a named YAML file) vs. implicit (the default `.semgrep.yml`)

Reply:
Hm this is a good point, an explicit named YAML file is probably better. Depends what we want to the standard AppSec rule authorship flow to be - is it a per repo directory of rules? Is it a company-specific semgrep-rules repo? Worth discussing, but I'd maybe lean towards the latter?